File: cpp/Problem40.cpp
/*
 * Problem: 40. Combination Sum II
 * URL: https://leetcode.com/problems/combination-sum-ii/
 * Difficulty: Medium
 * Tags: Array, Permutation, Backtracking
 */

#include <algorithm>
#include <vector>
#include <unordered_map>

using namespace std;

class Solution {
public:
	vector<vector<int>> combinationSum2(vector<int>& candidates, int target) {
		vector<vector<int>> results;
		vector<int> permutations;
		sort(candidates.begin(), candidates.end());
		backtrack(results, candidates, permutations, target, 0);
		return results;
	}

private:
	void backtrack(vector<vector<int>>& results, vector<int>& candidates, vector<int>& permutations, int target, int index) {
		if (target == 0) {
			results.push_back(permutations);
			return;
		}
		for (int i = index; i < candidates.size() && candidates[i] <= target; i++) {
			if (i > index && candidates[i] == candidates[i - 1]) {
				continue;
			}
			permutations.push_back(candidates[i]);
			backtrack(results, candidates, permutations, target - candidates[i], i + 1);
			permutations.pop_back();
		}
	}
};

File: python/Problem40.py
# Problem: 40. Combination Sum II
# URL: https://leetcode.com/problems/combination-sum-ii/
# Difficulty: Medium
# Tags: Array, Permutation, Backtracking

from typing import List

class Solution:
  def combinationSum2(self, candidates: List[int], target: int) -> List[List[int]]:
    results: List[List[int]] = []
    permutations: List[int] = []
    def backtrack(target: int, index: int):
      if target == 0:
        results.append(permutations[:])
        return
      for i in range(index, len(candidates)):
        if candidates[i] > target:
          break
        if i > index and candidates[i] == candidates[i - 1]:
          continue
        permutations.append(candidates[i])
        backtrack(target - candidates[i], i + 1)
        permutations.pop()
    candidates = sorted(candidates)
    backtrack(target, 0)
    return results